proto-lens-protoc/src/Data/ProtoLens/Compiler/Plugin.hs

```diff
@@ -0,0 +1,125 @@
+-- Copyright 2016 Google Inc. All Rights Reserved.
+--
+-- Use of this source code is governed by a BSD-style
+-- license that can be found in the LICENSE file or at
+-- https://developers.google.com/open-source/licenses/bsd
+--
+-- Code for writing protocol compiler plugins.
+
+{-# LANGUAGE OverloadedStrings #-}
+module Data.ProtoLens.Compiler.Plugin
+    ( ProtoFileName
+    , ProtoFile(..)
+    , analyzeProtoFiles
+    , buildEnv
+    , outputFilePath
+    , moduleName
+    , moduleNameStr
+    ) where
+
+import Data.Char (toUpper)
+import Data.List (foldl', intercalate)
+import qualified Data.Map.Strict as Map
+import Data.Map.Strict (Map, unions, (!))
+import Data.Monoid ((<>))
+import qualified Data.Text as T
+import Data.Text (Text)
+import Language.Haskell.Exts.Syntax (ModuleName(..), Name(..), QName(..))
+import Lens.Family2
+import Bootstrap.Proto.Google.Protobuf.Descriptor
+    (FileDescriptorProto, name, dependency, publicDependency)
+import System.FilePath (dropExtension, splitDirectories)
+
+
+import Data.ProtoLens.Compiler.Definitions
+
+-- | The filename of an input .proto file.
+type ProtoFileName = Text
+
+data ProtoFile = ProtoFile
+  { descriptor :: FileDescriptorProto
+  , haskellModule :: ModuleName
+  , definitions :: Env Name
+  , exports :: [ProtoFileName]
+  , exportedEnv :: Env QName
+  }
+
+-- Given a list of FileDescriptorProtos, collect information about each file
+-- into a map of 'ProtoFile's keyed by 'ProtoFileName'.
+analyzeProtoFiles :: Text -> [FileDescriptorProto] -> Map ProtoFileName ProtoFile
+analyzeProtoFiles modulePrefix files =
+    Map.fromList [ (f ^. name, ingestFile f) | f <- files ]
+  where
+    filesByName = Map.fromList [(f ^. name, f) | f <- files]
+    moduleNames = fmap (moduleName modulePrefix) filesByName
+    -- The definitions in each input proto file, indexed by filename.
+    definitionsByName = fmap collectDefinitions filesByName
+    -- The exports from each .proto file (including any "public import"
+    -- dependencies), as they appear to other modules that are importing them;
+    -- i.e., qualified by module name.
+    exportsByName = transitiveExports files
+    localExports = Map.intersectionWith qualifyEnv moduleNames definitionsByName
+    exportedEnvs = fmap (\es -> unions [localExports ! e | e <- es]) exportsByName
+
+    ingestFile f = ProtoFile
+        { descriptor = f
+        , haskellModule = moduleNames ! n
+        , definitions = definitionsByName ! n
+        , exports = exportsByName ! n
+        , exportedEnv = exportedEnvs ! n
+        }
+      where
+        n = f ^. name
+
+buildEnv :: [Text] -> Map ProtoFileName ProtoFile -> Env QName
```

Comment:
Maybe name this something a little more specific, like `collectEnvFromDeps`?  Up to you.

Also:  pass `[ProtoFileName]` as first arg instead of `[Text]`?